### ISSUE TITLE
Filter when using setTransformer searches on collection prior to the transformation.

### ISSUE THREAD

**Issue by prYmea** [yajra/laravel-datatables#248]
I am using a transformer on the datatables like below:

```
return \Datatables::of($company->events)
            ->setTransformer(DTEventTransformer::class)
            ->make(true);
```

However when I do a search on the datatable I am getting an error (see below) due to the fact that the CollectionEngine 'filtering()' method is doing a filter on the `$company->events` instead of the transformed data.

```
exception 'ErrorException' with message 'mb_strtolower() expects parameter 1 to be string, array given' in /laravel/framework/src/Illuminate/Support/Str.php:169
```

`$company->events`:
```
...
    "type": {
        "id": 1,
        "name": "created",
        "description": "Created event"
    },
...
```

After transformation:
```
...
    "type": "created",
...
```

As you can see, `type` is being passed as the array which is causing the error.

**Comment by yajra** [comment]
@prYmea, I rarely use the collection engine to this extent. But maybe you should use query builder approach instead like below?

```php
return \Datatables::of($company->events()->select())
            ->setTransformer(DTEventTransformer::class)
            ->make(true);
```

**Comment by prYmea** [comment]
@yajra That wouldn't make a difference as I have relations that are being pulled (hence the array `type` above).

It would be great if the filtering is done on the transformed output if a transformer is use.

**Comment by yajra** [comment]
Were you able to give it a shot? CollectionEngine behaves differently against QueryBuilderEngine.

For filtering, it is currently done before transforming the output. Your case maybe doable when using Collection but for query builder, I think it's not doable.

Lastly, the error you provide points to the framework. Can you provide an error stack the relates to the package? Or if possible, please send a PR to fix this? Thanks

**Comment by bart** [comment]
I do have a similar question. I'm using the query builder to build a really dynamic and complicate query with a lot of joins and selects inside. Now I need some additional information per row so I've added a transformer which generates these additional information and attaches it to every row. But how can I filter through all of these "attached" data? If I add a filter() method to the ```return Datatable::of($searchResults)``` it is done BEFORE the transformation has been applied, isn't it?

**Comment by yajra** [comment]
@bart YES, we do filtering before transforming or rendering the data for all the engines we have atm. (Eloquent, QueryBuilder, Collection). 

**For query builder, filter will only works if it works on normal SQL queries.** This is one reason why I added a Collection engine to help solve issues where columns are not part of the query. Although, I must admit that it's not yet full featured and needs an enhancements. 

> Can I drop results from inside the transformer? 

I think not if you need it, just **NOTE** that when using dataTables, **if data is not part of the query, then disable ordering & searching.**

**Comment by bart** [comment]
Thanks @yajra for your quick reply. I found a non-perfect but okay solution for my case. Instead of removing items I call the null() method which replaces the item by an empty one.

After the make() method has been called, I iterate through the getData() data array and remove empty objects. The one and only problem with this approach is, that you will have a lower result than expected in case at least one item has been removed.
